### Import Data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)

In [2]:
# Load Data
df = pd.read_csv('heart_disease.csv')

In [3]:
# Quick Preview:
print("shape:  ", df.shape)
print("Number of Rows:  " , df.shape[0])
print("Number of Columns :  ", df.shape[1])
df.head()

shape:   (4238, 16)
Number of Rows:   4238
Number of Columns :   16


Gender  age      education  currentSmoker  cigsPerDay  BPMeds  \
0    Male   39   postgraduate              0         0.0     0.0   
1  Female   46  primaryschool              0         0.0     0.0   
2    Male   48     uneducated              1        20.0     0.0   
3  Female   61       graduate              1        30.0     0.0   
4  Female   46       graduate              1        23.0     0.0   

  prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0              no             0         0    195.0  106.0   70.0  26.97   
1              no             0         0    250.0  121.0   81.0  28.73   
2              no             0         0    245.0  127.5   80.0  25.34   
3              no             1         0    225.0  150.0   95.0  28.58   
4              no             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [4]:
display(df.dtypes)

Gender              object
age                  int64
education           object
currentSmoker        int64
cigsPerDay         float64
BPMeds             float64
prevalentStroke     object
prevalentHyp         int64
diabetes             int64
totChol            float64
sysBP              float64
diaBP              float64
BMI                float64
heartRate          float64
glucose            float64
Heart_ stroke       object
dtype: object

### Cleaning Data

In [5]:
df.describe()

age  currentSmoker   cigsPerDay       BPMeds  prevalentHyp  \
count  4238.000000    4238.000000  4209.000000  4185.000000   4238.000000   
mean     49.584946       0.494101     9.003089     0.029630      0.310524   
std       8.572160       0.500024    11.920094     0.169584      0.462763   
min      32.000000       0.000000     0.000000     0.000000      0.000000   
25%      42.000000       0.000000     0.000000     0.000000      0.000000   
50%      49.000000       0.000000     0.000000     0.000000      0.000000   
75%      56.000000       1.000000    20.000000     0.000000      1.000000   
max      70.000000       1.000000    70.000000     1.000000      1.000000   

          diabetes      totChol        sysBP        diaBP          BMI  \
count  4238.000000  4188.000000  4238.000000  4238.000000  4219.000000   
mean      0.025720   236.721585   132.352407    82.893464    25.802008   
std       0.158316    44.590334    22.038097    11.910850     4.080111   
min       0.000000   107.000000    83.500000    48.000000    15.540000   
25%       0.000000   206.000000   117.000000    75.000000    23.070000   
50%       0.000000   234.000000   128.000000    82.000000    25.400000   
75%       0.000000   263.000000   144.000000    89.875000    28.040000   
max       1.000000   696.000000   295.000000   142.500000    56.800000   

         heartRate      glucose  
count  4237.000000  3850.000000  
mean     75.878924    81.966753  
std      12.026596    23.959998  
min      44.000000    40.000000  
25%      68.000000    71.000000  
50%      75.000000    78.000000  
75%      83.000000    87.000000  
max     143.000000   394.000000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4238 entries, 0 to 4237
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Gender           4238 non-null   object 
 1   age              4238 non-null   int64  
 2   education        4133 non-null   object 
 3   currentSmoker    4238 non-null   int64  
 4   cigsPerDay       4209 non-null   float64
 5   BPMeds           4185 non-null   float64
 6   prevalentStroke  4238 non-null   object 
 7   prevalentHyp     4238 non-null   int64  
 8   diabetes         4238 non-null   int64  
 9   totChol          4188 non-null   float64
 10  sysBP            4238 non-null   float64
 11  diaBP            4238 non-null   float64
 12  BMI              4219 non-null   float64
 13  heartRate        4237 non-null   float64
 14  glucose          3850 non-null   float64
 15  Heart_ stroke    4238 non-null   object 
dtypes: float64(8), int64(4), object(4)
memory usage: 529.9+ KB


In [7]:
# Number of Nulls in all features
df.isnull().sum().sort_values(ascending=False) 

glucose            388
education          105
BPMeds              53
totChol             50
cigsPerDay          29
BMI                 19
heartRate            1
Gender               0
age                  0
currentSmoker        0
prevalentStroke      0
prevalentHyp         0
diabetes             0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

In [8]:
# Null % descending for features
((df.isnull().sum() / df.value_counts().sum()) *100).sort_values(ascending=False)

glucose            10.612691
education           2.871991
BPMeds              1.449672
totChol             1.367615
cigsPerDay          0.793217
BMI                 0.519694
heartRate           0.027352
Gender              0.000000
age                 0.000000
currentSmoker       0.000000
prevalentStroke     0.000000
prevalentHyp        0.000000
diabetes            0.000000
sysBP               0.000000
diaBP               0.000000
Heart_ stroke       0.000000
dtype: float64

In [9]:
df['education'].value_counts()

education
uneducated       1720
primaryschool    1253
graduate          687
postgraduate      473
Name: count, dtype: int64

1- education feature

In [10]:
# Fill Null Values with Mode (uneducated)
df['education'].fillna(df['education'].mode()[0], inplace=True)

In [11]:
df['education'].value_counts()

education
uneducated       1825
primaryschool    1253
graduate          687
postgraduate      473
Name: count, dtype: int64

In [12]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
BPMeds              53
totChol             50
cigsPerDay          29
BMI                 19
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
prevalentStroke      0
prevalentHyp         0
diabetes             0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

In [13]:
df.head()

Gender  age      education  currentSmoker  cigsPerDay  BPMeds  \
0    Male   39   postgraduate              0         0.0     0.0   
1  Female   46  primaryschool              0         0.0     0.0   
2    Male   48     uneducated              1        20.0     0.0   
3  Female   61       graduate              1        30.0     0.0   
4  Female   46       graduate              1        23.0     0.0   

  prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0              no             0         0    195.0  106.0   70.0  26.97   
1              no             0         0    250.0  121.0   81.0  28.73   
2              no             0         0    245.0  127.5   80.0  25.34   
3              no             1         0    225.0  150.0   95.0  28.58   
4              no             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

cigsPerDay --> Ciggrates Number per Day

In [14]:
# Fill Null Values with median (0)
df['cigsPerDay'].fillna(df['cigsPerDay'].median(), inplace=True)

In [15]:
# convert data type to int
df['cigsPerDay'] = df['cigsPerDay'].astype('int')

In [16]:
df.head()

Gender  age      education  currentSmoker  cigsPerDay  BPMeds  \
0    Male   39   postgraduate              0           0     0.0   
1  Female   46  primaryschool              0           0     0.0   
2    Male   48     uneducated              1          20     0.0   
3  Female   61       graduate              1          30     0.0   
4  Female   46       graduate              1          23     0.0   

  prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0              no             0         0    195.0  106.0   70.0  26.97   
1              no             0         0    250.0  121.0   81.0  28.73   
2              no             0         0    245.0  127.5   80.0  25.34   
3              no             1         0    225.0  150.0   95.0  28.58   
4              no             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [17]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
BPMeds              53
totChol             50
BMI                 19
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
prevalentStroke      0
prevalentHyp         0
diabetes             0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

In [18]:
df.head()

Gender  age      education  currentSmoker  cigsPerDay  BPMeds  \
0    Male   39   postgraduate              0           0     0.0   
1  Female   46  primaryschool              0           0     0.0   
2    Male   48     uneducated              1          20     0.0   
3  Female   61       graduate              1          30     0.0   
4  Female   46       graduate              1          23     0.0   

  prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0              no             0         0    195.0  106.0   70.0  26.97   
1              no             0         0    250.0  121.0   81.0  28.73   
2              no             0         0    245.0  127.5   80.0  25.34   
3              no             1         0    225.0  150.0   95.0  28.58   
4              no             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [19]:
# Relace vales to Smoker and Not Smoker
df['currentSmoker'] = df['currentSmoker'].replace({0: 'Not Somker', 1: 'Smoker'})

In [20]:
df.head()

Gender  age      education currentSmoker  cigsPerDay  BPMeds  \
0    Male   39   postgraduate    Not Somker           0     0.0   
1  Female   46  primaryschool    Not Somker           0     0.0   
2    Male   48     uneducated        Smoker          20     0.0   
3  Female   61       graduate        Smoker          30     0.0   
4  Female   46       graduate        Smoker          23     0.0   

  prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0              no             0         0    195.0  106.0   70.0  26.97   
1              no             0         0    250.0  121.0   81.0  28.73   
2              no             0         0    245.0  127.5   80.0  25.34   
3              no             1         0    225.0  150.0   95.0  28.58   
4              no             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [21]:
df['currentSmoker'].value_counts()

currentSmoker
Not Somker    2144
Smoker        2094
Name: count, dtype: int64

BPMeds --> Blood Pressure Medications.

In [22]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
BPMeds              53
totChol             50
BMI                 19
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
prevalentStroke      0
prevalentHyp         0
diabetes             0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

In [23]:
df['BPMeds'].value_counts()

BPMeds
0.0    4061
1.0     124
Name: count, dtype: int64

In [24]:
# Replace Null Values with mode
df['BPMeds'].fillna(df['BPMeds'].mode()[0], inplace=True)

In [25]:
# Convert Data Type to int  
df['BPMeds'] = df['BPMeds'].astype('int')

In [26]:
df['BPMeds'].value_counts()

BPMeds
0    4114
1     124
Name: count, dtype: int64

In [27]:
# Replace Values to NO Meds and On Meds
df['BPMeds'] = df['BPMeds'].replace({0: 'NO Meds', 1:'On Meds'})

In [28]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0              no             0         0    195.0  106.0   70.0  26.97   
1              no             0         0    250.0  121.0   81.0  28.73   
2              no             0         0    245.0  127.5   80.0  25.34   
3              no             1         0    225.0  150.0   95.0  28.58   
4              no             0         0    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [29]:
df['BPMeds'].value_counts()

BPMeds
NO Meds    4114
On Meds     124
Name: count, dtype: int64

In [30]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
totChol             50
BMI                 19
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
BPMeds               0
prevalentStroke      0
prevalentHyp         0
diabetes             0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

prevalentStroke --> has had a stroke before

In [31]:
# Replace Values to make it start with Capital letter
df['prevalentStroke'] = df['prevalentStroke'].replace({'no' : 'No', 'yes' : 'Yes'}) 

In [32]:
df['prevalentStroke'].value_counts()

prevalentStroke
No     4213
Yes      25
Name: count, dtype: int64

prevalentHyp --> person has hypertension (high blood pressure)

In [33]:
# Replace 0,1 to Values Yes and No
df['prevalentHyp'] = df['prevalentHyp'].replace({0: 'No', 1:'Yes'})

diabetes --> person has diabetes

In [34]:
# Replace 0,1 to Values Yes and No
df['diabetes'] = df['diabetes'].replace({0: 'No', 1:'Yes'})

In [35]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [36]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
totChol             50
BMI                 19
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
BPMeds               0
prevalentStroke      0
prevalentHyp         0
diabetes             0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

In [37]:
# Replace the Null Values with Median (234)
df['totChol'].fillna(df['totChol'].median(), inplace = True)

In [38]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
BMI                 19
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
BPMeds               0
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol              0
sysBP                0
diaBP                0
Heart_ stroke        0
dtype: int64

In [39]:
# Replace Null Vales with Median(25.4)
df['BMI'].fillna(df['BMI'].median(), inplace=True)

In [40]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [41]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
heartRate            1
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
BPMeds               0
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol              0
sysBP                0
diaBP                0
BMI                  0
Heart_ stroke        0
dtype: int64

In [42]:
# Replace Null Vales with Median(75)
df['heartRate'].fillna(int(df['heartRate'].median()), inplace=True)

In [43]:
df.isnull().sum().sort_values(ascending=False)

glucose            388
Gender               0
age                  0
education            0
currentSmoker        0
cigsPerDay           0
BPMeds               0
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol              0
sysBP                0
diaBP                0
BMI                  0
heartRate            0
Heart_ stroke        0
dtype: int64

In [44]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [45]:
# Replace Null Vales with Median(78)
df['glucose'].fillna(df['glucose'].median(), inplace=True)

In [46]:
df.isna().sum()

Gender             0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
Heart_ stroke      0
dtype: int64

In [47]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           yes  
4       85.0     85.0            No

In [48]:
df['Heart_ stroke'].value_counts()

Heart_ stroke
No     3594
yes     644
Name: count, dtype: int64

In [49]:
# Replace yes with Yes
df['Heart_ stroke'] = df['Heart_ stroke'].replace({'yes':'Yes'})

In [50]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           Yes  
4       85.0     85.0            No

In [51]:
# Remove duplicate rows

before = len(df)
df = df.drop_duplicates().reset_index(drop = True)
after = len(df)
print(f"Dropped {before - after} duplicate rows . current rows : {after}")

Dropped 0 duplicate rows . current rows : 4238


In [52]:
# Strip leading &trailing Spaces from all text columns
for col in df.select_dtypes(include = 'object'):
  df[col] = df[col].str.strip()

print("Trimmed whitespace in text columns.")
df.head(5)

Trimmed whitespace in text columns.


Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke  
0       80.0     77.0            No  
1       95.0     76.0            No  
2       75.0     70.0            No  
3       65.0    103.0           Yes  
4       85.0     85.0            No

In [53]:
display(df.dtypes)

Gender              object
age                  int64
education           object
currentSmoker       object
cigsPerDay           int32
BPMeds              object
prevalentStroke     object
prevalentHyp        object
diabetes            object
totChol            float64
sysBP              float64
diaBP              float64
BMI                float64
heartRate          float64
glucose            float64
Heart_ stroke       object
dtype: object

In [54]:
for col in df.select_dtypes(include='object').columns:
    print(df[col].value_counts())

Gender
Female    2419
Male      1819
Name: count, dtype: int64
education
uneducated       1825
primaryschool    1253
graduate          687
postgraduate      473
Name: count, dtype: int64
currentSmoker
Not Somker    2144
Smoker        2094
Name: count, dtype: int64
BPMeds
NO Meds    4114
On Meds     124
Name: count, dtype: int64
prevalentStroke
No     4213
Yes      25
Name: count, dtype: int64
prevalentHyp
No     2922
Yes    1316
Name: count, dtype: int64
diabetes
No     4129
Yes     109
Name: count, dtype: int64
Heart_ stroke
No     3594
Yes     644
Name: count, dtype: int64


### Feature Engineering

In [55]:
# Add Column age_group
bins = [ 30, 40, 50, 60, 70, 80]
labels = ['30-39', '40-49','50-59','60-69','70+']
df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels, right=False)

In [56]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke age_group  
0       80.0     77.0            No     30-39  
1       95.0     76.0            No     40-49  
2       75.0     70.0            No     40-49  
3       65.0    103.0           Yes     60-69  
4       85.0     85.0            No     40-49

In [57]:
df['age_group'].value_counts()

age_group
40-49    1660
50-59    1333
60-69     688
30-39     555
70+         2
Name: count, dtype: int64

In [58]:
# Categorize the BMI into 5 Categories
bins = [0, 18.5, 25, 30, 40, 100]
labels = ['Under_Weight', 'Normal', 'Over_Weight', 'Obese', 'Extremely Obese']
df['BMI_Cat'] = pd.cut(df['BMI'], bins=bins, labels=labels, right=False)

In [59]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke age_group      BMI_Cat  
0       80.0     77.0            No     30-39  Over_Weight  
1       95.0     76.0            No     40-49  Over_Weight  
2       75.0     70.0            No     40-49  Over_Weight  
3       65.0    103.0           Yes     60-69  Over_Weight  
4       85.0     85.0            No     40-49       Normal

In [60]:
df['BMI_Cat'].value_counts()

BMI_Cat
Normal             1869
Over_Weight        1773
Obese               513
Under_Weight         57
Extremely Obese      26
Name: count, dtype: int64

In [61]:
# Make Ciggrates smoked per day into Ranges
bins = [0, 1, 6, 11, 16, 21, 26, 31, 36, 41, 100]
labels = ['Not_Smoker', '1-5', '6-10', '11-15', '16-20', '21-25', '26-30', '31-35', '36-40', '+41']
df['Cigs_Range'] = pd.cut(df['cigsPerDay'], bins=bins, labels=labels, right=False)

In [62]:
df['Cigs_Range'] = df['Cigs_Range'].astype('str')

In [63]:
df['Cigs_Range'].value_counts().sort_values(ascending=False)

Cigs_Range
Not_Smoker    2173
16-20          754
1-5            315
6-10           314
11-15          223
26-30          218
36-40           81
+41             77
21-25           61
31-35           22
Name: count, dtype: int64

In [64]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke age_group      BMI_Cat  Cigs_Range  
0       80.0     77.0            No     30-39  Over_Weight  Not_Smoker  
1       95.0     76.0            No     40-49  Over_Weight  Not_Smoker  
2       75.0     70.0            No     40-49  Over_Weight       16-20  
3       65.0    103.0           Yes     60-69  Over_Weight       26-30  
4       85.0     85.0            No     40-49       Normal       21-25

In [65]:
# Categorize Data into Cholesterol Levels
bins = [0, 200, 240, 750]
labels = ['Optimal', 'Intermediate', 'High']
df['Chol_Cat'] = pd.cut(df['totChol'], bins=bins, labels=labels, right=False)

In [66]:
df['Chol_Cat'].value_counts()

Chol_Cat
High            1881
Intermediate    1528
Optimal          829
Name: count, dtype: int64

In [67]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  sysBP  diaBP    BMI  \
0              No           No       No    195.0  106.0   70.0  26.97   
1              No           No       No    250.0  121.0   81.0  28.73   
2              No           No       No    245.0  127.5   80.0  25.34   
3              No          Yes       No    225.0  150.0   95.0  28.58   
4              No           No       No    285.0  130.0   84.0  23.10   

   heartRate  glucose Heart_ stroke age_group      BMI_Cat  Cigs_Range  \
0       80.0     77.0            No     30-39  Over_Weight  Not_Smoker   
1       95.0     76.0            No     40-49  Over_Weight  Not_Smoker   
2       75.0     70.0            No     40-49  Over_Weight       16-20   
3       65.0    103.0           Yes     60-69  Over_Weight       26-30   
4       85.0     85.0            No     40-49       Normal       21-25   

       Chol_Cat  
0       Optimal  
1          High  
2          High  
3  Intermediate  
4          High

### sysBP --> Systolic Blood Pressure
### diaBP --> Diastolic Blood Pressure

In [68]:
# Categorize Data into Systolic Blood Pressure Levels
bins = [0, 120, 130, 140, 160, 180, 300]
labels = ["Normal", "Elevated", "High-Normal", "Hypertension Stage 1", "Hypertension Stage 2", "Hypertensive Crisis"]
df['Sys_Cat'] = pd.cut(df['sysBP'], bins=bins, labels=labels, right=False)

In [69]:
# Categorize Data into Diastolic Blood Pressure Levels 
bins = [0, 80, 85, 90, 100, 110, 180]
labels = ["Normal", "Elevated" ,"High-Normal", "Hypertension Stage 1", "Hypertension Stage 2", "Hypertensive Crisis"]
df['Dia_Cat'] = pd.cut(df['diaBP'], bins=bins, labels=labels, right=False)

In [70]:
df['Sys_Cat'].value_counts()

Sys_Cat
Normal                  1279
Elevated                 947
Hypertension Stage 1     805
High-Normal              729
Hypertension Stage 2     308
Hypertensive Crisis      170
Name: count, dtype: int64

In [71]:
df['Dia_Cat'].value_counts()

Dia_Cat
Normal                  1666
Elevated                 864
Hypertension Stage 1     695
High-Normal              648
Hypertension Stage 2     252
Hypertensive Crisis      113
Name: count, dtype: int64

In [72]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  ...    BMI  heartRate  \
0              No           No       No    195.0  ...  26.97       80.0   
1              No           No       No    250.0  ...  28.73       95.0   
2              No           No       No    245.0  ...  25.34       75.0   
3              No          Yes       No    225.0  ...  28.58       65.0   
4              No           No       No    285.0  ...  23.10       85.0   

   glucose  Heart_ stroke  age_group      BMI_Cat  Cigs_Range      Chol_Cat  \
0     77.0             No      30-39  Over_Weight  Not_Smoker       Optimal   
1     76.0             No      40-49  Over_Weight  Not_Smoker          High   
2     70.0             No      40-49  Over_Weight       16-20          High   
3    103.0            Yes      60-69  Over_Weight       26-30  Intermediate   
4     85.0             No      40-49       Normal       21-25          High   

                Sys_Cat               Dia_Cat  
0                Normal                Normal  
1              Elevated              Elevated  
2              Elevated              Elevated  
3  Hypertension Stage 1  Hypertension Stage 1  
4           High-Normal              Elevated  

[5 rows x 22 columns]

In [73]:
risk_order = ["Normal", "Elevated", "High-Normal", "Hypertension Stage 1", "Hypertension Stage 2", "Hypertensive Crisis"]
def max_risk(row):
    sys_idx = risk_order.index(row['Sys_Cat'])
    dia_idx = risk_order.index(row['Dia_Cat'])
    return risk_order[max(sys_idx, dia_idx)]
df['BP_Cat'] = df.apply(max_risk, axis=1)

In [74]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  ...  heartRate  glucose  \
0              No           No       No    195.0  ...       80.0     77.0   
1              No           No       No    250.0  ...       95.0     76.0   
2              No           No       No    245.0  ...       75.0     70.0   
3              No          Yes       No    225.0  ...       65.0    103.0   
4              No           No       No    285.0  ...       85.0     85.0   

   Heart_ stroke  age_group      BMI_Cat  Cigs_Range      Chol_Cat  \
0             No      30-39  Over_Weight  Not_Smoker       Optimal   
1             No      40-49  Over_Weight  Not_Smoker          High   
2             No      40-49  Over_Weight       16-20          High   
3            Yes      60-69  Over_Weight       26-30  Intermediate   
4             No      40-49       Normal       21-25          High   

                Sys_Cat               Dia_Cat                BP_Cat  
0                Normal                Normal                Normal  
1              Elevated              Elevated              Elevated  
2              Elevated              Elevated              Elevated  
3  Hypertension Stage 1  Hypertension Stage 1  Hypertension Stage 1  
4           High-Normal              Elevated           High-Normal  

[5 rows x 23 columns]

In [75]:
# Categorize Data into Heart Rate Levels 

# Bradycardia(Slow) ---- Tachycardia(Fast)  
bins = [0, 60, 101, 150]
labels = ['Bradycardia', 'Normal', 'Tachycardia'] 
df['HeartRate_Cat'] = pd.cut(df['heartRate'], bins=bins, labels=labels, right=False)

In [76]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  ...  glucose  Heart_ stroke  \
0              No           No       No    195.0  ...     77.0             No   
1              No           No       No    250.0  ...     76.0             No   
2              No           No       No    245.0  ...     70.0             No   
3              No          Yes       No    225.0  ...    103.0            Yes   
4              No           No       No    285.0  ...     85.0             No   

   age_group      BMI_Cat  Cigs_Range      Chol_Cat               Sys_Cat  \
0      30-39  Over_Weight  Not_Smoker       Optimal                Normal   
1      40-49  Over_Weight  Not_Smoker          High              Elevated   
2      40-49  Over_Weight       16-20          High              Elevated   
3      60-69  Over_Weight       26-30  Intermediate  Hypertension Stage 1   
4      40-49       Normal       21-25          High           High-Normal   

                Dia_Cat                BP_Cat HeartRate_Cat  
0                Normal                Normal        Normal  
1              Elevated              Elevated        Normal  
2              Elevated              Elevated        Normal  
3  Hypertension Stage 1  Hypertension Stage 1        Normal  
4              Elevated           High-Normal        Normal  

[5 rows x 24 columns]

In [77]:
df['HeartRate_Cat'].value_counts()

HeartRate_Cat
Normal         3927
Bradycardia     206
Tachycardia     105
Name: count, dtype: int64

In [78]:
# Categorize Data into Glucose Levels 
bins = [0, 70, 100, 126, 400]
labels = ['Hypoglycemia', 'Normal', 'Prediabetes', 'Diabetes'] 
df['Glucose_Cat'] = pd.cut(df['glucose'], bins=bins, labels=labels, right=False)

In [79]:
df['Glucose_Cat'].value_counts() 

Glucose_Cat
Normal          3131
Hypoglycemia     746
Prediabetes      275
Diabetes          86
Name: count, dtype: int64

In [80]:
df.head()

Gender  age      education currentSmoker  cigsPerDay   BPMeds  \
0    Male   39   postgraduate    Not Somker           0  NO Meds   
1  Female   46  primaryschool    Not Somker           0  NO Meds   
2    Male   48     uneducated        Smoker          20  NO Meds   
3  Female   61       graduate        Smoker          30  NO Meds   
4  Female   46       graduate        Smoker          23  NO Meds   

  prevalentStroke prevalentHyp diabetes  totChol  ...  Heart_ stroke  \
0              No           No       No    195.0  ...             No   
1              No           No       No    250.0  ...             No   
2              No           No       No    245.0  ...             No   
3              No          Yes       No    225.0  ...            Yes   
4              No           No       No    285.0  ...             No   

   age_group      BMI_Cat  Cigs_Range      Chol_Cat               Sys_Cat  \
0      30-39  Over_Weight  Not_Smoker       Optimal                Normal   
1      40-49  Over_Weight  Not_Smoker          High              Elevated   
2      40-49  Over_Weight       16-20          High              Elevated   
3      60-69  Over_Weight       26-30  Intermediate  Hypertension Stage 1   
4      40-49       Normal       21-25          High           High-Normal   

                Dia_Cat                BP_Cat HeartRate_Cat  Glucose_Cat  
0                Normal                Normal        Normal       Normal  
1              Elevated              Elevated        Normal       Normal  
2              Elevated              Elevated        Normal       Normal  
3  Hypertension Stage 1  Hypertension Stage 1        Normal  Prediabetes  
4              Elevated           High-Normal        Normal       Normal  

[5 rows x 25 columns]

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4238 entries, 0 to 4237
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   Gender           4238 non-null   object  
 1   age              4238 non-null   int64   
 2   education        4238 non-null   object  
 3   currentSmoker    4238 non-null   object  
 4   cigsPerDay       4238 non-null   int32   
 5   BPMeds           4238 non-null   object  
 6   prevalentStroke  4238 non-null   object  
 7   prevalentHyp     4238 non-null   object  
 8   diabetes         4238 non-null   object  
 9   totChol          4238 non-null   float64 
 10  sysBP            4238 non-null   float64 
 11  diaBP            4238 non-null   float64 
 12  BMI              4238 non-null   float64 
 13  heartRate        4238 non-null   float64 
 14  glucose          4238 non-null   float64 
 15  Heart_ stroke    4238 non-null   object  
 16  age_group        4238 non-null   category


In [82]:
output_path = 'heart_disease_cleaned.csv'
df.to_csv(output_path, index=False)

### Detect Outliers (UpNoraml Values)

In [83]:
# outliers by IQR for column 'glucose'


# Calculate IQR
q1 = df['glucose'].quantile(0.25)
q3 = df['glucose'].quantile(0.75)
iqr = q3 - q1

# Define bounds
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr


outliers = df[(df['glucose'] < lower_bound) | (df['glucose'] > upper_bound)]
print(f"New Number of raws in 'glucose': {len(outliers)}")



New Number of raws in 'glucose': 262


### glucose (Blood Sugar): This column has the most outliers (262 cases).

***Observation*** : The maximum value is 394 mg/dL, while the upper bound for "normal" data distribution in this set is around 111 mg/dL.

***Insight*** : These high values likely represent patients with uncontrolled diabetes. Since this is a heart disease study, these are scientifically important values, not necessarily errors.

In [84]:
# outliers by IQR for systolic_BP (sysBP)

# Calculate IQR
q1 = df['sysBP'].quantile(0.25)
q3 = df['sysBP'].quantile(0.75)
iqr = q3 - q1

# Define bounds
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

outliers = df[(df['sysBP'] < lower_bound) | (df['sysBP'] > upper_bound)]
print(f"New Number of raws in 'sysBP': {len(outliers)}")

New Number of raws in 'sysBP': 126


### sysBP (Systolic Blood Pressure): 126 outliers

***Observation*** : The maximum is 295 mmHg, which is extremely high (hypertensive crisis level).

***Insight*** : Like glucose, these are critical high-risk cases for heart disease prediction.

In [85]:
# outliers by IQR for 'totChol'

# Calculate IQR
q1 = df['totChol'].quantile(0.25)
q3 = df['totChol'].quantile(0.75)
iqr = q3 - q1

# Define bounds
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

outliers = df[(df['totChol'] < lower_bound) | (df['totChol'] > upper_bound)]
print(f"New Number of raws in 'totChol': {len(outliers)}")

New Number of raws in 'totChol': 57


### totChol (Total Cholesterol): 57 outliers

***Observation*** : The maximum is 696 mg/dL, which is very high compared to the upper bound of 349.

In [86]:
# outliers by IQR for BMI

# Calculate IQR
q1 = df['BMI'].quantile(0.25)
q3 = df['BMI'].quantile(0.75)
iqr = q3 - q1

lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

outliers = df[(df['BMI'] < lower_bound) | (df['BMI'] > upper_bound)]
print(f"New Number of raws in 'BMI': {len(outliers)}")

New Number of raws in 'BMI': 97


### BMI: 97 outliers
***Observation*** : Max is 56.8, indicating severe obesity cases.

In [87]:
# outliers by IQR for 'cigsPerDay'

# Calculate IQR
q1 = df['cigsPerDay'].quantile(0.25)
q3 = df['cigsPerDay'].quantile(0.75)
iqr = q3 - q1

lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

outliers = df[(df['cigsPerDay'] < lower_bound) | (df['cigsPerDay'] > upper_bound)]
print(f"New Number of raws in 'cigsPerDay': {len(outliers)}")

New Number of raws in 'cigsPerDay': 12


### cigsPerDay: Only 12 outliers
 ***Observation*** : Most smokers in the set smoke fewer than 50 cigarettes a day, but a few heavy smokers reach up to 70.